File: app/lib/engine/failsafe.dart
/// Dart port of packages/engine/src/core/failsafe.ts.
/// Bezpečnostne kritické: keď tvár vypadne, nesmie prejsť odkrytý frame.
/// Hysterézia: clona sa zapne rýchlo, vypne opatrne.

class FailSafe {
  final int coverAfter;
  final int recoverAfter;

  bool _covered = true;
  int _missStreak = 0;
  int _hitStreak = 0;

  FailSafe({this.coverAfter = 2, this.recoverAfter = 5});

  /// Zaznamená detekciu pre jeden frame; vráti, či má byť obraz zaclonený.
  bool update(bool faceDetected) {
    if (faceDetected) {
      _hitStreak++;
      _missStreak = 0;
      if (_covered && _hitStreak >= recoverAfter) {
        _covered = false;
      }
    } else {
      _missStreak++;
      _hitStreak = 0;
      if (!_covered && _missStreak >= coverAfter) {
        _covered = true;
      }
    }
    return _covered;
  }

  bool get isCovered => _covered;

  void reset() {
    _covered = true;
    _missStreak = 0;
    _hitStreak = 0;
  }
}

File: packages/engine/src/core/failsafe.ts
/**
 * Fail-safe stavový automat.
 *
 * Bezpečnostne kritické: keď detekcia tváre vypadne, NESMIE prejsť žiadny
 * odkrytý frame. Naopak, krátke výpadky detekcie (1–2 framy) nesmú blikať
 * clonou. Preto hysterézia: clona sa zapne rýchlo, vypne opatrne.
 */

export interface FailSafeOptions {
  /** Po koľkých po sebe idúcich frameoch bez tváre zaclonení (rýchlo). */
  coverAfter: number;
  /** Koľko po sebe idúcich frameov s tvárou treba na odclonenie (opatrne). */
  recoverAfter: number;
}

export const DEFAULT_FAILSAFE: FailSafeOptions = {
  coverAfter: 2,
  recoverAfter: 5,
};

/**
 * Sleduje históriu detekcie a rozhoduje, či má byť obraz zaclonený.
 * Štartuje v zaclonenom stave (covered = true) — radšej skry, kým si istý.
 */
export class FailSafe {
  private readonly opts: FailSafeOptions;
  private covered = true;
  private missStreak = 0;
  private hitStreak = 0;

  constructor(options: Partial<FailSafeOptions> = {}) {
    this.opts = { ...DEFAULT_FAILSAFE, ...options };
  }

  /**
   * Zaznamená výsledok detekcie pre jeden frame a vráti, či má byť
   * obraz zaclonený celoplošnou clonou.
   */
  update(faceDetected: boolean): boolean {
    if (faceDetected) {
      this.hitStreak++;
      this.missStreak = 0;
      if (this.covered && this.hitStreak >= this.opts.recoverAfter) {
        this.covered = false;
      }
    } else {
      this.missStreak++;
      this.hitStreak = 0;
      if (!this.covered && this.missStreak >= this.opts.coverAfter) {
        this.covered = true;
      }
    }
    return this.covered;
  }

  /** Aktuálny stav bez zmeny. */
  isCovered(): boolean {
    return this.covered;
  }

  /** Reset do bezpečného (zaclonneného) stavu. */
  reset(): void {
    this.covered = true;
    this.missStreak = 0;
    this.hitStreak = 0;
  }
}
